### ISSUE TITLE
NTLM support broken

### ISSUE THREAD

**Issue by avf** [Alamofire/Alamofire#721]
Hi,
in Alamofire 1.2.2 I could make NTLM requests with the following code:

        let user = "" //insert user here
        let password = "" //insert pw
        let url = "" //insert URL
        
        let credential = NSURLCredential(user: user, password: password, persistence: .None)
        
        Alamofire.request(.GET, url)
            .authenticate(usingCredential: credential)
            .responseString { request, response, data, error in
                println("\(response), \(data), \(error)")
        }

However, in Alamofire 1.3.1 this only returns a 401 Unauthorized. I didn't change anything in my own code, I only dropped the newer framework into my project.

I've done a bit of investigation, and in the current master, `- URLSession:task:didReceiveChallenge:completionHandler:` is not called on the `SessionDelegate`. The commit that broke this seems to be b69959b8ad129057a9f4f313a0ccc7772902ee65. With the removal of `respondsToSelector`, the above method isn't called anymore. However, I don't understand why. Could someone please look into it?

**Comment by cnoon** [comment]
Is there any possible way you could provide us sample code that we can actual run to help us debug the issue? We already authentication challenge tests that are passing as expected.

**Comment by cnoon** [comment]
Hi @avf, I need your help to debug this. I don't have a way to test NTLM authentication. If you could provide more information that would be really helpful. Would you be able to provide actual sample code that I could run? If you're worried about obfuscation, would you be willing to work with me through email or Slack?

We have tests around this for basic and digest auth that work just fine with that code change. Please get back to me asap. I'd like to get this tied off before the Alamofire 2.0 release.

**Comment by avf** [comment]
Hi @cnoon, sorry for not replying earlier. Unfortunately, I'm not working on the project where I needed the NTLM support anymore and I also don't have access to the NTLM server anymore. I googled a bit and tried to find an easy way to test the protocol, but as you might now it's difficult to find an open webserver where you can just send some test requests to. Best I could find is [this](http://stackoverflow.com/questions/6509278/authentication-test-servers), which would require to setup a test apache server. Sorry I can't be of more help!

**Comment by cnoon** [comment]
Thanks for getting back to me @avf. Unfortunately, since we don't have any way to test this, I'm going to have to close this out for now. If someone runs across this issue again, we can re-open or debug at that time.

Cheers. 🍻

**Comment by mubashirjavaid** [comment]
I am facing this issue at the moment while trying to access sharepoint rest api, under ssl and ntlm

**Comment by cnoon** [comment]
Would you be willing to help me debug this issue @mubashirjavaid? It may take a bit of time to figure out the issue. I may need you to try some different branches to try out some different approaches. Also, would you be able to provide me with enough sample code that I could debug the problem myself. I'd be happy to work with you in a non-public facing way such as through Slack or email.

**Comment by mubashirjavaid** [comment]
Yes I can do it, We can have session on skype together. My sky ID is mubashir.javaid

**Comment by cnoon** [comment]
Got some awesome info for debugging this from @mubashirjavaid! Once I have a chance to dig through the issue I'll report back what I find. Hopefully I'll have a fix.

**Comment by cnoon** [comment]
Okay...got to the bottom of the issue. Thank you so much @mubashirjavaid for helping me debug the problem. I wouldn't have been able to figure out what the root cause was without your help! 🍻

The issue was in fact the commit b69959b. That was a very bad change on my part. I have reverted that change in a9480269 and made another bug fix in 6ac4ce17. I was able to verify that a9480269 does in fact fix the NTLM issue. It also fixes some other unwanted behaviors.

I'll be pushing out 2.0.1 here shortly that will include this fix. Thanks everyone for helping get to the bottom of this issue. Alamofire is much better off for it!
